### PR TITLE
Utils.popen_read: Nonexistent program should fail

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Ensure that `$CHILD_STATUS.success?` is false when the command does not exist.

### CODE REVIEW

**Comment by sjackman** [comment]
Currently `$CHILD_STATUS.success?` is `true` after executing a non-existant program.
```
❯❯❯ brew irb <<<'Utils.popen_read("nonexistant"); $CHILD_STATUS.success?'
==> Interactive Homebrew Shell
Example commands available with: brew irb --examples
Switch to inspect mode.
Utils.popen_read("nonexistant"); $CHILD_STATUS.success?
true
```

With this patch, it's false, as expected.
```
❯❯❯ brew irb <<<'Utils.popen_read("nonexistant"); $CHILD_STATUS.success?'
==> Interactive Homebrew Shell
Example commands available with: brew irb --examples
Switch to inspect mode.
Utils.popen_read("nonexistant"); $CHILD_STATUS.success?
false
```

**Comment by sjackman** [comment]
I tried to write a test, but ran into some trouble. It looks as though the test harness is catching the non-existant file before `Utils.popen_read` does.
```ruby
❯❯❯ brew tests --only=utils/popen                         popen
Failures:

  1) Utils::popen_read fails when the command does not exist
     Failure/Error:
       Dir.foreach(@path) {|e|
         next if e == '.' || e == '..'
         if with_directory
           result << self.class.new(File.join(@path, e))
         else
           result << self.class.new(e)
         end
       }
     
     Errno::ENOENT:
       No such file or directory @ dir_initialize - /tmp/homebrew-tests-20171201-56677-bq281e/prefix/Library/Taps/homebrew/homebrew-core/Formula
     # ./test/spec_helper.rb:115:in `map'
     # ./test/spec_helper.rb:115:in `block (2 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `block (2 levels) in <top (required)>'
```

**Comment by MikeMcQuaid** [comment]
> It looks as though the test harness is catching the non-existant file before Utils.popen_read does.

That doesn't seem to be the exception here because that's not the directory it's trying to create. You should be able to catch `Errno::ENOENT` and inspect it if that's the problem.

File: Library/Homebrew/utils/popen.rb [hunk 1]
```diff
@@ -14,7 +14,12 @@ def self.popen(args, mode, options = {})
         yield pipe
       else
         options[:err] ||= :close unless ENV["HOMEBREW_STDERR"]
-        exec(*args, options)
+        begin
+          exec(*args, options)
+        rescue SystemCallError
+          $stderr.puts "brew: command not found: #{args[0]}" unless options[:err] == :close
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/popen.rb]: This message doesn't seem to match the exception; presumably there are other `SystemCallError`s?

**Comment by ilovezfs** [comment]
s/existant/existent/

File: Library/Homebrew/utils/popen.rb (hunk 1 above)
Comment (sjackman) [reply Library/Homebrew/utils/popen.rb]: > You should be able to catch Errno::ENOENT and inspect it if that's the problem.

Thanks for the tip. Fixed.

**Comment by sjackman** [comment]
`exit` should have been `exit!`. `exit` incorrectly invokes the `at_exit` handlers. The child process should not invoke the `at_exit` handlers. That should be left to the parent process.

**Comment by MikeMcQuaid** [comment]
`brew style` is 🔴

**Comment by MikeMcQuaid** [comment]
Thanks again @sjackman!

**Comment by sjackman** [comment]
Thanks again for merging, Mike!

**Pull request merged**